### PR TITLE
C++: Fix small PrintAST and PrintIR issue

### PR DESCRIPTION
**Pull request by jketema**
* PrintIR is currently doing the wrong thing for `OverloadedArrayExpr` add a test and fix
* PrintIR currently does the wrong thing if a block has two outgoing edges to the same successor block. At present this is not an issue, but it will be once https://github.com/github/codeql/pull/18253 is merged. I tested this on top of that PR, but since merging that PR is somewhat blocked at present, I would like to get this in preemptively.

#### All query authors

- [ ] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [ ] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.
- [ ] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [ ] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [ ] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/PrintIR.qll [hunk 1]
```diff
@@ -427,11 +427,18 @@ query predicate edges(PrintableIRBlock pred, PrintableIRBlock succ, string key,
     succBlock = succ.getBlock() and
     predBlock.getSuccessor(kind) = succBlock and
     (
-      (
-        key = "semmle.label" and
+      key = "semmle.label" and
+      exists(string kinds |
+        kinds =
+          concat(EdgeKind k |
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/PrintIR.qll:433]: If this is changed to a `strictcount` you could remove line 428, right?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/PrintIR.qll:433]: In the `key = "semmle.order"` case? That's bound by `getSuccessorIndex` anyway. Although, I do see that you need to rewrite the `exists` slightly to get rid of the resulting cartesian product with `kind`.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/PrintIR.qll:433]: > In the `key = "semmle.order"` case? That's bound by `getSuccessorIndex` anyway.

Ah, you're right.

So are you thinking of something like:
```ql
query predicate edges(PrintableIRBlock pred, PrintableIRBlock succ, string key, string value) {
  exists(IRBlock predBlock, IRBlock succBlock |
    predBlock = pred.getBlock() and
    succBlock = succ.getBlock() and
    (
      key = "semmle.label" and
      exists(string kinds |
        kinds =
          strictconcat(EdgeKind k |
            predBlock.getSuccessor(k) = succBlock
          |
            k.toString(), "|" order by k.toString()
          )
      |
        if predBlock.getBackEdgeSuccessor(_) = succBlock
        then value = kinds + " (back edge)"
        else value = kinds
      )
      or
      key = "semmle.order" and
      value = getSuccessorIndex(predBlock, succBlock).toString()
    )
  )
}
```

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/PrintIR.qll:433]: Exactly 👍

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/PrintIR.qll]: Done in f9bfd969a99cd7b2238e921a09d97088fd4e09eb

**Pull request merged**